tactics/hints.ml

```diff
@@ -916,19 +902,27 @@ let fresh_global_or_constr env sigma poly cr =
       (c, Univ.ContextSet.empty)
     end
 
-let make_resolves env sigma flags info ~check ~poly ?name cr =
+let make_resolves env sigma (eapply, hnf) info ~check ~poly ?name cr =
   let c, ctx = fresh_global_or_constr env sigma poly cr in
   let cty = Retyping.get_type_of env sigma c in
   let try_apply f =
-    try Some (f (c, cty, ctx)) with Failure _ -> None in
+    try
+      let (hd, hint) as ans = f (c, cty, ctx) in
+      let () = match hint.code.obj with
+      | ERes_pf _ -> if not eapply then failwith "make_apply_entry"
+      | _ -> ()
+      in
+      Some ans
+    with Failure _ -> None
```

Comment:
Should we avoid raising just to be caught?
```suggestion
    try
      let (hd, hint) as ans = f (c, cty, ctx) in
      match hint.code.obj with
      | ERes_pf _ -> None
      | _ -> Some ans 
    with Failure _ -> None
```

Reply:
I thought about that but the other part of the code still relies on raising exceptions, so I didn't want to change this behaviour too far...

Reply:
(Your change is wrong btw, it needs to check eapply on the relevant branch.)

Reply:
Done.